client/keys.go

```diff
@@ -88,30 +88,31 @@ func EndorsementKeyFromNvIndex(rw io.ReadWriter, idx uint32) (*Key, error) {
 // function will only work on a GCE VM. Unlike AttestationKeyRSA, this key uses
 // the Endorsement Hierarchy and its template loaded from GceAKTemplateNVIndexRSA.
 func GceAttestationKeyRSA(rw io.ReadWriter) (*Key, error) {
-	akRsa, err := EndorsementKeyFromNvIndex(rw, GceAKTemplateNVIndexRSA)
+	gceAkRsa, err := CachedKeyFromNvIndex(rw, tpm2.HandleEndorsement, GceAKTemplateNVIndexRSA, GceAKRSAHandle)
 	if err != nil {
 		return nil, err
 	}
-	if err := akRsa.trySetCertificateFromNvram(GceAKCertNVIndexRSA); err != nil {
-		akRsa.Close()
+	if err := gceAkRsa.trySetCertificateFromNvram(GceAKCertNVIndexRSA); err != nil {
+		gceAkRsa.Close()
 		return nil, err
 	}
-	return akRsa, nil
+	return gceAkRsa, nil
 }
 
 // GceAttestationKeyECC generates and loads the GCE ECC AK. Note that this
 // function will only work on a GCE VM. Unlike AttestationKeyECC, this key uses
 // the Endorsement Hierarchy and its template loaded from GceAKTemplateNVIndexECC.
+// The key will be persisted to the GceAKECCHandle
 func GceAttestationKeyECC(rw io.ReadWriter) (*Key, error) {
-	akEcc, err := EndorsementKeyFromNvIndex(rw, GceAKTemplateNVIndexECC)
+	gceAkEcc, err := CachedKeyFromNvIndex(rw, tpm2.HandleEndorsement, GceAKTemplateNVIndexECC, GceAKECCHandle)
 	if err != nil {
 		return nil, err
 	}
-	if err := akEcc.trySetCertificateFromNvram(GceAKCertNVIndexECC); err != nil {
-		akEcc.Close()
+	if err := gceAkEcc.trySetCertificateFromNvram(GceAKCertNVIndexECC); err != nil {
+		gceAkEcc.Close()
 		return nil, err
 	}
-	return akEcc, nil
+	return gceAkEcc, nil
 }
```

Comment:
After thinking I think it makes sense to cached the key, because it will also cached the Cert, which will otherwise get read from NVdata every time. I reverted the key change.

Question is how to cache it efficiently and easy to read, i added customized key fetcher in launcher, let me know what you think.

Reply:
was trying to keep the FetchAttestation signature, now just save it to agent struct.